Improve the documentation of `Display` and `FromStr`, and their interactions

joshtriplett opened the pull request:
In particular:
- `Display` is not necessarily lossless
- The output of `Display` might not be parseable by `FromStr`, and might
  not produce the same value if it is.
- Calling `.parse()` on the output of `Display` is usually a mistake
  unless a type's documented output and input formats match.
- The input formats accepted by `FromStr` depend on the type.

This documentation adds no API surface area and makes no guarantees about stability. To the best of my knowledge, everything it says is already established to be true. As such, I don't think it needs an FCP.

joshtriplett committed:
Improve the documentation of `Display` and `FromStr`, and their interactions

In particular:
- `Display` is not necessarily lossless
- The output of `Display` might not be parseable by `FromStr`, and might
  not produce the same value if it is.
- Calling `.parse()` on the output of `Display` is usually a mistake
  unless a type's documented output and input formats match.
- The input formats accepted by `FromStr` depend on the type.

BurntSushi commented:
I think I generally agree here, but I do think making `Display` and `FromStr` match is _good practice_ where possible. Maybe we can tweak the language here to accommodate that?

joshtriplett committed:
Add some more description of interactions between `Display` and `FromStr`

BurntSushi commented:
I think there might be a misunderstanding? I'm saying that _if_ your type provides both `Display` and `FromStr` impls, _then_ it's good practice to make them compatible with one another. If your type lacks one or both of the impls, then I agree, the advice doesn't apply. (As is the case for `Path`, `OsStr`, `ByteStr` and so on.)

I agree about the delimiters and what not. But I think that's a different can of worms. :-)

ChrisDenton commented:
> If your type lacks one or both of the impls

Aside: Lacking only `Display` is strongly discouraged by the docs: https://doc.rust-lang.org/std/string/trait.ToString.html

> This trait is automatically implemented for any type which implements the [Display](https://doc.rust-lang.org/std/fmt/trait.Display.html) trait. As such, ToString shouldn’t be implemented directly: [Display](https://doc.rust-lang.org/std/fmt/trait.Display.html) should be implemented instead, and you get the ToString implementation for free.

Personally I don't love the tight coupling between `Display` and `ToString` but there's no use complaining about it now and I'll admit it can be convenient.

joshtriplett commented:
I added a note that "Having both `Display` and `FromStr` implementations where the result of `Display` cannot be parsed with `FromStr` may surprise users.", along with the qualifier that that *doesn't* mean the result of parsing will have exactly the same value as the input.

BurntSushi commented:
Thanks! I'm happy with that.

joshtriplett commented:
> Thanks! I'm happy with that.

(Is that an r= or would you prefer someone else to review? :) )

BurntSushi commented on library/core/src/fmt/mod.rs:
```diff
@@ -931,6 +931,15 @@ pub use macros::Debug;
+/// possible to parse, and even if it is, the result of parsing might not exactly match the original
+/// value. Calling `.parse()` on the output from `Display` is usually a mistake, unless the type has
+/// provided and documented additional guarantees about its `Display` and `FromStr` implementations.
+///
```
I think this still reads to me to be a bit to pessimistic on coordinating `Display` and `FromStr` impls. And similarly for below. I liked your phrasing about surprises in the comments, but I'm not sure if that has made it in here yet?

Like if your type has _both_ a `Display` and a `FromStr` impl, I do think it is idiomatic for those to be congruent with one another. It may not be lossless and so on, but like, I think this is in direct contrast to what you say here:

> Calling `.parse()` on the output from `Display` is usually a mistake

This reads as very confusing to me, because this is not my perspective on ecosystem and std convention. (I use "convention" purposely here to accentuate that I am not promulgating a guarantee. I agree we don't have a guarantees here and _shouldn't_ have guarantees, because there are exceptions. But I think the norm is for `Display` and `FromStr` to be in alignment.)

joshtriplett replied:
@BurntSushi
> I liked your phrasing about surprises in the comments, but I'm not sure if that has made it in here yet?

I thought I had pushed it, but the push failed. I've now pushed it.

> It may not be lossless and so on, but like, I think this is in direct contrast to what you say here:
> > `Calling .parse() on the output from Display is usually a mistake`
>
> This reads as very confusing to me, because this is not my perspective on ecosystem and std convention

If the sentence ended there with a full stop, I would agree with you. The point is that you shouldn't do so *unless you know the type actually supports it*; *assuming* it without checking the type is a mistake.

joshtriplett replied:
I *think* I've resolved this, but feel free to unresolve if not.

BurntSushi commented:
I did a bit more of a careful review. I think the phrasing here is still a bit too pessimistic for my liking.

joshtriplett commented:
@BurntSushi I *think* I've addressed your concern here; please let me know if the current version doesn't address your concern.

the8472 commented:
One example in the standard library where a type implements both Display and FromStr but doesn't roundtrip is [SocketAddrV6](https://doc.rust-lang.org/nightly/std/net/struct.SocketAddrV6.html).

Amanieu commented:
Reading through this thread, it seems that the discussion fundamentally boils down to 2 questions:

* On a type that implements both `Display` and `FromStr`, can a user to assume that the `FromStr` implementation can parse the output of the `Display` implementation?

* If yes then is the `Display` implementation guaranteed to be complete? This implies that displaying and parsing is able to fully preserve the value.

Obviously a type may "override" these guarantees with documentation. We only need to specify the default expectations in the absence of specific documentation to the contrary.

---

I believe that the answer to the first question is yes: the default expectation should be that `FromStr` is able to parse the format that is output by a `Display` impl. However the second question is trickier and I think it would be reasonable to warn users against expecting round-tripping unless specifically promised by type documentation.

ijackson commented:
Having `Display` just display things seems justifiable.  But `ToString::to_string` being lossy is much more surprising.  Indeed it's arguably a violation of the albeit unofficial [Rust API Guidelines](https://rust-lang.github.io/api-guidelines/naming.html#c-conv).

Maybe in the 2027 edition we can have edition-specific traits (I'm looking at you, `std::error::Error`) somehow and then we can have a version of `.to_string()` that only works if the type roundtrips through `String`.

Or we could rename `ToString::to_string` to `ToDisplayString::to_display_string` and have a new `ToFaithfulString::to_string` implemented for only types that opt in somehow.

Both of these would mean a semantic difference between `format("{x}")` and `x.to_string_sensibly()` but that seems fair to me.

ijackson commented:
> Having `Display` just display things seems justifiable. But `ToString::to_string` being lossy is much more surprising. Indeed it's arguably a violation of the albeit unofficial [Rust API Guidelines](https://rust-lang.github.io/api-guidelines/naming.html#c-conv).

So err maybe we could add some docs to `ToString` in the meantime, where we apologise for this beartrap?

jdahlstrom commented:
> But ToString::to_string being lossy is much more surprising. Indeed it's arguably a violation of the albeit unofficial [Rust API Guidelines](https://rust-lang.github.io/api-guidelines/naming.html#c-conv).

I don't see how. The guidelines don't say or recommend that `to_` conversions should be injective.

---

> In a type that implements both Display and FromStr, can a user to assume that the FromStr implementation can parse the output of the Display implementation?

My 2c is that `from_str(&to_string(x)) == Ok(x)` should be guaranteed for at least all primitive types. Currently `FromStr for f32` documents the grammar that it accepts (including ±Inf and NaN), but `Display for f32` *doesn't* specify what it outputs, although in practice I expect all `f32` bit patterns to be round-trippable (modulo NaN payload bits). Integer types don't specify anything about how they format/parse strings, but I presume that they "obviously" round-trip.

joshtriplett committed:
`FromStr`: Rework explanation of `FromStr`/`Display` round-tripping

- Drop the phrasing "usually a mistake".
- Mention that `Display` may not be lossless.
- Drop a misplaced parenthetical about round-tripping that didn't fit
  the paragraph it was in.

joshtriplett committed:
`Display`: Rework explanation of `FromStr`/`Display` round-tripping

- Drop "usually a mistake"
- Add phrasing from `FromStr` about round-tripping, and about how the
  inability to round-trip may surprise users.

joshtriplett commented:
Reworked based on discussion with @Amanieu to tone down the "usually a mistake" phrasing, and focus on caveats and guidance about potential user confusion.

bors commented:
:pushpin: Commit 742014e7e3576f3053e7c63fc930c4d5b5b7a477 has been approved by `Amanieu`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Amanieu 742014e7e3576f3053e7c63fc930c4d5b5b7a477 -->
<!-- homu: {"type":"Approved","sha":"742014e7e3576f3053e7c63fc930c4d5b5b7a477","approver":"Amanieu","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
